pkg/metrics/prometheus.go

```diff
@@ -16,60 +16,185 @@
+// SetGauge sets a gauge metric with a given name, value, and labels.
+func (p *prometheusCollector) SetGauge(_ context.Context, name string, value float64, labels ...string) error {
+	name = fmt.Sprintf("%s_%s", p.name, name)
+	if _, ok := p.gauges[name]; !ok {
+		return fmt.Errorf("gauge '%s' not found", name)
 	}
+	p.gauges[name].WithLabelValues(labels...).Set(value)
+	return nil
 }
 
-// NewCounter creates a new counter metric
-func (p *prom) NewCounter(name string) error {
-	if _, ok := p.counters[name]; ok {
-		return fmt.Errorf("counter %s already exists", name)
+// AddHistogram increments a histogram metric with a given name, value, and labels.
+func (p *prometheusCollector) AddHistogram(_ context.Context, name string, value float64, labels ...string) error {
+	name = fmt.Sprintf("%s_%s", p.name, name)
+	if _, ok := p.histograms[name]; !ok {
+		return fmt.Errorf("histogram '%s' not found", name)
+	}
+	p.histograms[name].WithLabelValues(labels...).Observe(value)
+	return nil
+}
+
+// AddCounter increments a counter metric with a given name, value, and labels.
+// It returns an error if the counter is not found.
+func (p *prometheusCollector) AddCounter(_ context.Context, name string, value float64, labels ...string) error {
+	name = fmt.Sprintf("%s_%s", p.name, name)
+	if _, ok := p.counters[name]; !ok {
+		return fmt.Errorf("counter '%s' not found", name)
 	}
-	counter := prometheus.NewCounter(
-		prometheus.CounterOpts{
-			Name: fmt.Sprintf("%s_total", name),
-			Help: fmt.Sprintf("The total number of %s function calls", name),
-		})
-	prometheus.MustRegister(counter)
-	p.counters[name] = counter
+	p.counters[name].WithLabelValues(labels...).Add(value)
 	return nil
 }
 
-// IncrementCounter increments a counter metric
-func (p *prom) IncrementCounter(funcName string) {
-	p.counters[funcName].Inc()
+// MetricsHandler returns a http.Handler for the prometheus metrics.
+func (p *prometheusCollector) MetricsHandler() http.Handler {
+	return promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{
+		EnableOpenMetrics: true,
+	})
 }
 
-// NewSummary creates a new summary metric
-func (p *prom) NewSummary(name string) error {
-	if _, ok := p.summary[name]; ok {
-		return fmt.Errorf("summary %s already exists", name)
+// NewPrometheus creates a new prometheusCollector with empty sync.Maps for histograms, gauges, and counters.
+func NewPrometheus(name string) MetricCollector {
+	name = fmt.Sprintf("%s_%s_function_duration_seconds", "guac", name)
+	functionDuration := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    name,
+			Help:    "Time spent executing functions.",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"function"},
+	)
+	p := &prometheusCollector{
+		histograms: make(map[string]*prometheus.HistogramVec),
+		gauges:     make(map[string]*prometheus.GaugeVec),
+		counters:   make(map[string]*prometheus.CounterVec),
+	}
+	p.histograms[name] = functionDuration
```

Comment:
the docstring states that this should be an empty histogram?

Reply:
I don't understand your comment. Could you please explain?